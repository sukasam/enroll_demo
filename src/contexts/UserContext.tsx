import { dismissAllErrorToasts } from "Components/shared/ShowToaster";
import { AddressType } from "Contexts/types/OrderContextTypes";
import { UserContextState, UserData } from "Contexts/types/UserContextTypes";
import { useRouter } from "next/router";
import {
    ReactElement,
    ReactNode,
    createContext,
    useCallback,
    useContext,
    useEffect,
    useMemo,
    useState
} from "react";
import useLoginHooks from "./hooks/LoginHooks";

const defaultState: UserContextState = {
    userToken: null,
    setUserToken: () => null,
    enrollerId: null,
    setEnrollerId: () => null,
    sponsorId: null,
    setSponsorId: () => null,
    userData: null,
    setUserData: () => null,
    enrollerFullName: null,
    setEnrollerFullName: () => null,
    sponsorFullName: null,
    setSponsorFullName: () => null,
    enrollerAllFullName: null,
    setEnrollerAllFullName: () => null,
    isLoggedIn: false,
    setIsLoggedIn: () => null,
    href: null,
    setHref: () => null,
    userLocale: null,
    setUserLocale: () => null,
    isUserContextLoading: true,
    setIsUserContextLoading: () => null,
    activeAccordionSection: null,
    setActiveAccordionSection: () => null,
    goToNextSection: () => null,
    setLoggedInData: () => null,
    mainAddress: null as AddressType | null, // Update type
    setMainAddress: () => null,
    deleteAuthCookie: () => null,
    resetBacktoDefault: () => null
};

const UserContext = createContext<UserContextState>(defaultState);

export function UserProvider({
    children,
    testValues
}: {
    children: ReactNode;
    testValues?: Record<string, string>;
}): ReactElement {
    const [userToken, setUserToken] = useState<string | null>(
        defaultState.userToken
    );
    const [enrollerId, setEnrollerId] = useState<string | null>(
        defaultState.enrollerId
    );
    const [sponsorId, setSponsorId] = useState<string | null>(
        defaultState.sponsorId
    );
    const [userData, setUserData] = useState<UserData | null>(
        defaultState.userData
    );
    const [enrollerFullName, setEnrollerFullName] = useState<string | null>(
        defaultState.enrollerFullName
    );
    const [sponsorFullName, setSponsorFullName] = useState<string | null>(
        defaultState.sponsorFullName
    );
    const [enrollerAllFullName, setEnrollerAllFullName] = useState<any | null>(
        defaultState.enrollerAllFullName
    );
    const [isLoggedIn, setIsLoggedIn] = useState<boolean>(
        defaultState.isLoggedIn
    );
    const [href, setHref] = useState<string | null>(defaultState.href);
    const [userLocale, setUserLocale] = useState<string | null>(
        defaultState.userLocale
    );
    const [isUserContextLoading, setIsUserContextLoading] = useState<boolean>(
        defaultState.isUserContextLoading
    );
    const [activeAccordionSection, setActiveAccordionSection] = useState<
        number | null
    >(defaultState.activeAccordionSection);
    const [mainAddress, setMainAddress] = useState<AddressType | null>(
        defaultState.mainAddress
    );

    const router = useRouter();
    const {
        getUserTokenFromCookie,
        setAuthCookie,
        deleteAuthCookie: _deleteAuthCookie,
        setLoggedInData: _setLoggedInData
    } = useLoginHooks();

    const resetBacktoDefault = useCallback(() => {
        setUserToken(defaultState.userToken);
        setEnrollerId(defaultState.enrollerId);
        setSponsorId(defaultState.sponsorId);
        setUserData(defaultState.userData);
        setEnrollerFullName(defaultState.enrollerFullName);
        setSponsorFullName(defaultState.sponsorFullName);
        setEnrollerAllFullName(defaultState.enrollerAllFullName);
        setIsLoggedIn(defaultState.isLoggedIn);
        setHref(defaultState.href);
        setUserLocale(defaultState.userLocale);
        setIsUserContextLoading(defaultState.isUserContextLoading);
        setActiveAccordionSection(defaultState.activeAccordionSection);
        setMainAddress(defaultState.mainAddress);
    }, [
        setUserToken,
        setEnrollerId,
        setSponsorId,
        setUserData,
        setEnrollerFullName,
        setSponsorFullName,
        setEnrollerAllFullName,
        setIsLoggedIn,
        setHref,
        setUserLocale,
        setIsUserContextLoading,
        setActiveAccordionSection,
        setMainAddress
    ]);

    const deleteAuthCookie = useCallback(() => {
        _deleteAuthCookie();
        resetBacktoDefault();
    }, [_deleteAuthCookie, resetBacktoDefault]);

    const setLoggedInData = useCallback(
        async (token: string): Promise<void> => {
            try {
                await _setLoggedInData({
                    userToken: token,
                    setUserData,
                    setHref,
                    setUserToken,
                    setEnrollerFullName,
                    setSponsorFullName,
                    setEnrollerId,
                    setSponsorId,
                    setMainAddress
                });
            } catch (error) {
                deleteAuthCookie();
                resetBacktoDefault();
                throw error;
            }
        },
        [
            _setLoggedInData,
            setUserData,
            setHref,
            setUserToken,
            setEnrollerFullName,
            setSponsorFullName,
            setEnrollerId,
            setSponsorId,
            setMainAddress,
            resetBacktoDefault
        ]
    );

    useEffect(() => {
        setIsLoggedIn(!!userToken);
        setIsUserContextLoading(false);
    }, [userToken]);

    const loadCachedUserData = async (): Promise<void> => {
        setIsUserContextLoading(true);
        const _userToken = getUserTokenFromCookie();
        setUserToken(_userToken);

        const cachedData = sessionStorage.getItem("userContext");
        if (cachedData) {
            try {
                const parsedData: UserContextState = JSON.parse(cachedData);

                setEnrollerId(parsedData.enrollerId);
                setSponsorId(parsedData.sponsorId);
                setUserData(parsedData.userData);
                setEnrollerFullName(parsedData.enrollerFullName);
                setSponsorFullName(parsedData.sponsorFullName);
                setEnrollerAllFullName(parsedData.enrollerAllFullName);
                setIsLoggedIn(!!_userToken);
                setHref(parsedData.href);
                setUserLocale(parsedData.userLocale);
                setMainAddress(parsedData.mainAddress);

                if (!parsedData.href && _userToken) {
                    await setLoggedInData(_userToken);
                }
            } catch (error) {
                console.error(
                    "Failed to parse user context data from sessionStorage:",
                    error
                );
            }
        }
        setIsUserContextLoading(false);
    };

    useEffect(() => {
        if (router.pathname !== "/logout" && router.pathname !== "/reset")
            loadCachedUserData();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const updateCachedUserData = useCallback((): void => {
        setAuthCookie(userToken);
        const dataToCache = {
            enrollerId,
            sponsorId,
            userData,
            enrollerFullName,
            sponsorFullName,
            enrollerAllFullName,
            isLoggedIn,
            href,
            userLocale,
            activeAccordionSection,
            mainAddress
        };
        sessionStorage.setItem("userContext", JSON.stringify(dataToCache));
    }, [
        setAuthCookie,
        userToken,
        enrollerId,
        sponsorId,
        userData,
        enrollerFullName,
        sponsorFullName,
        enrollerAllFullName,
        isLoggedIn,
        href,
        userLocale,
        activeAccordionSection,
        mainAddress
    ]);

    useEffect(() => {
        updateCachedUserData();
    }, [updateCachedUserData]);

    const goToNextSection = useCallback(() => {
        dismissAllErrorToasts();
        setActiveAccordionSection(prevSection =>
            prevSection !== null ? prevSection + 1 : 1
        );
    }, [setActiveAccordionSection]);

    const value = useMemo(
        () => ({
            userToken,
            setUserToken,
            enrollerId,
            setEnrollerId,
            sponsorId,
            setSponsorId,
            userData,
            setUserData,
            enrollerFullName,
            setEnrollerFullName,
            sponsorFullName,
            setSponsorFullName,
            enrollerAllFullName,
            setEnrollerAllFullName,
            isLoggedIn,
            setIsLoggedIn,
            href,
            setHref,
            userLocale,
            setUserLocale,
            isUserContextLoading,
            setIsUserContextLoading,
            activeAccordionSection,
            setActiveAccordionSection,
            goToNextSection,
            setLoggedInData,
            mainAddress,
            setMainAddress,
            deleteAuthCookie,
            resetBacktoDefault
        }),
        [
            userToken,
            enrollerId,
            sponsorId,
            userData,
            enrollerFullName,
            sponsorFullName,
            enrollerAllFullName,
            isLoggedIn,
            href,
            userLocale,
            isUserContextLoading,
            activeAccordionSection,
            goToNextSection,
            setLoggedInData,
            mainAddress,
            setMainAddress,
            deleteAuthCookie,
            resetBacktoDefault
        ]
    );

    const providerValue = useMemo(
        () => ({
            ...value,
            ...testValues
        }),
        [value, testValues]
    );

    return (
        <UserContext.Provider value={providerValue}>
            {children}
        </UserContext.Provider>
    );
}

export const useUser = (): UserContextState => useContext(UserContext);
