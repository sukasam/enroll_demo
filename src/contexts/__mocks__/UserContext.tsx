import { AddressType } from "Contexts/types/OrderContextTypes";

export const useUser = jest.fn().mockReturnValue({
    userToken: null,
    setUserToken: jest.fn(),
    enrollerId: null,
    setEnrollerId: jest.fn(),
    sponsorId: null,
    setSponsorId: jest.fn(),
    userData: null,
    setUserData: jest.fn(),
    enrollerFullName: null,
    setEnrollerFullName: jest.fn(),
    sponsorFullName: null,
    setSponsorFullName: jest.fn(),
    isLoggedIn: false,
    setIsLoggedIn: jest.fn(),
    href: null,
    setHref: jest.fn(),
    userLocale: null,
    setUserLocale: jest.fn(),
    isUserContextLoading: false, // almost all test need this to be false vs just the loading test scenario
    setIsUserContextLoading: jest.fn(),
    activeAccordionSection: null,
    setActiveAccordionSection: jest.fn(),
    goToNextSection: jest.fn(),
    setLoggedInData: jest.fn(),
    mainAddress: null as AddressType | null,
    setMainAddress: jest.fn(),
    deleteAuthCookie: jest.fn(),
    resetBacktoDefault: jest.fn()
});

export default useUser;
