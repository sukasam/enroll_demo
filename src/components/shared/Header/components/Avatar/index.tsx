import { Avatar as MuiAvatar, SxProps } from "@mui/material";
import { useUser } from "Contexts/UserContext";

function Avatar(): JSX.Element {
    const { userData } = useUser();
    const stringAvatar = (name: string): { sx: SxProps; children: string } => {
        const nameParts = name.split(" ");
        let initials = nameParts[0][0];

        if (nameParts.length > 1 && nameParts[1]) {
            initials += nameParts[1][0];
        }

        return {
            sx: {
                bgcolor: "#75a4d3",
                width: "24px",
                height: "24px",
                fontSize: "12px",
                marginRight: "8px"
            },
            children: initials
        };
    };

    const avatarProps = stringAvatar(userData?.fullName || "");
    return <MuiAvatar sx={avatarProps.sx}>{avatarProps.children}</MuiAvatar>;
}

export default Avatar;
