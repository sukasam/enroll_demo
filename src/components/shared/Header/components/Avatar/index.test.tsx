import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { useUser } from "Contexts/UserContext";
import Avatar from "./index";

// Mock the useUser hook
jest.mock("Contexts/UserContext", () => ({
    useUser: jest.fn()
}));

describe("Avatar", () => {
    it("renders correctly with full name", () => {
        (useUser as jest.Mock).mockReturnValue({
            userData: { fullName: "John Doe" }
        });

        const { container } = render(<Avatar />);
        expect(screen.getByText("JD")).toBeInTheDocument();
        expect(container.firstChild).toMatchSnapshot();
    });

    it("renders correctly with single name", () => {
        (useUser as jest.Mock).mockReturnValue({
            userData: { fullName: "John" }
        });

        render(<Avatar />);
        expect(screen.getByText("J")).toBeInTheDocument();
    });

    it("renders default avatar icon with empty name", () => {
        (useUser as jest.Mock).mockReturnValue({
            userData: { fullName: "" }
        });

        render(<Avatar />);
        expect(screen.getByTestId("PersonIcon")).toBeInTheDocument();
    });

    it("renders default avatar icon with undefined userData", () => {
        (useUser as jest.Mock).mockReturnValue({
            userData: undefined
        });

        render(<Avatar />);
        expect(screen.getByTestId("PersonIcon")).toBeInTheDocument();
    });

    it("applies correct styles", () => {
        (useUser as jest.Mock).mockReturnValue({
            userData: { fullName: "John Doe" }
        });

        render(<Avatar />);
        const avatar = screen.getByText("JD");
        expect(avatar).toHaveStyle({
            backgroundColor: "#75a4d3",
            width: "24px",
            height: "24px",
            fontSize: "12px",
            marginRight: "8px"
        });
    });
});
