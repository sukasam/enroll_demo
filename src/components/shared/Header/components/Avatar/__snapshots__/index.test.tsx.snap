// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly with full name 1`] = `
<div
  class=MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-XXXXX-MuiAvatar-root
>
  JD
</div>
`;
